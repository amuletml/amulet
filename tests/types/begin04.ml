class monad 'm begin
  val pure : 'a -> 'm 'a
  val (>>=) : 'm 'a -> ('a -> 'm 'b) -> 'm 'b
end

type identity 'a = Identity of 'a

instance monad identity begin
  let pure = Identity
  let (Identity x) >>= f = f x
end

instance monad list begin
  let pure x = [x]
  let xs >>= f = [ x | with x <- xs, with x <- f x ]
end

(* can't mix and match monads *)
let test = do
  let! x = Identity 1
  let! y = [2]
  pure 2
end
