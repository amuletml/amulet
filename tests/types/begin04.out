begin04.ml[21:3 ..21:15]: error (E2001)
   │ 
21 │   with y <- [2]
   │   ^^^^^^^^^^^^^
  Couldn't match actual type list int
    with the type expected by the context, identity 'b
